### ISSUE TITLE
[java] DU Anomaly

### ISSUE THREAD

**Issue by davidgoate** [pmd/pmd#1107]
Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core].

**Affects PMD Version:** PMD 6.0.1
**Rule:** [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis) DU-Anomaly

**Description:** DU-Anomaly on DU - Anomaly: A recently defined variable is undefined.
It's unclear to me why this complains.

**Code Sample demonstrating the issue:**

```java
public class Sample {

    private static final LocalTime ONE_PAST_MIDNIGHT = MIDNIGHT.plusMinutes(1);

    public ZonedDateTime sample(final String date, final ZonedDateTime paymentTimeStamp) {
        final ZonedDateTime startOfDayOnInceptionDate = zonedDateTimeAtOnePastMidnight(date, paymentTimeStamp.getZone());
        final ZonedDateTime startOfDayOnPaymentDate = zonedDateTimeAtOnePastMidnight(paymentTimeStamp.toLocalDate(), paymentTimeStamp.getZone());
        if (startOfDayOnInceptionDate.isAfter(paymentTimeStamp)) {
            return startOfDayOnInceptionDate;
        } else if (paymentTimeStamp.isAfter(startOfDayOnPaymentDate)) {
            return paymentTimeStamp;
        } else {
            return startOfDayOnPaymentDate;
        }
    }

    public ZonedDateTime zonedDateTimeAtOnePastMidnight(final String date, final ZoneId zoneId) {
        return zonedDateTimeAtOnePastMidnight(parseStringWithIsoDateFormat(date), zoneId);
    }

    private ZonedDateTime zonedDateTimeAtOnePastMidnight(final LocalDate localDate, final ZoneId zoneId) {
        return ZonedDateTime.of(localDate, ONE_PAST_MIDNIGHT, zoneId);
    }
}

```

Result:
`n/a:7:	Found 'DU'-anomaly for variable 'startOfDayOnPaymentDate' (lines '7'-'15').`

Complains about: `final ZonedDateTime startOfDayOnPaymentDate = zonedDateTimeAtOnePastMidnight(paymentTimeStamp.toLocalDate(), paymentTimeStamp.getZone());` but i don't quite see why given it's used in the `else if` and `else` and is `final` so isn't undefined at any point.

**Running PMD through:** Maven 3.9.0

**Comment by michaelhoefer** [comment]
@davidgoate and all - this looks correct to me based on what the rule is checking.  The confusing thing is the message.  The "problem" here is, you define a variable up front and in one path at least you never use it.  

"Undefinition" of a variable just means it no longer exists...because the scope where it was defined (the method body in this case) closed out.  You could reorg the code to avoid it and technically be more correct/optimized.  Since you are using multiple return statements you can also lose the else statements. 

```
final ZonedDateTime startOfDayOnInceptionDate = zonedDateTimeAtOnePastMidnight(date, paymentTimeStamp.getZone());
if (startOfDayOnInceptionDate.isAfter(paymentTimeStamp)) {
    return startOfDayOnInceptionDate;
}
final ZonedDateTime startOfDayOnPaymentDate =
    zonedDateTimeAtOnePastMidnight(paymentTimeStamp.toLocalDate(),
    paymentTimeStamp.getZone());
if (paymentTimeStamp.isAfter(startOfDayOnPaymentDate)) {
    return paymentTimeStamp;
}
return startOfDayOnPaymentDate;
```

**Comment by adangel** [comment]
The rule [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis) has been deprecated in favour of [UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment), which is available with PMD 6.26.0.

Therefore this issue wont be fixed.

See also #2647.
